src/PowerShellEditorServices/Console/UnixConsoleOperations.cs

```diff
@@ -11,54 +11,176 @@ internal class UnixConsoleOperations : IConsoleOperations
 
-        private static readonly ManualResetEventSlim _waitHandle = new ManualResetEventSlim();
+        private static readonly ManualResetEventSlim s_waitHandle = new ManualResetEventSlim();
 
-        private SemaphoreSlim _readKeyHandle = new SemaphoreSlim(1, 1);
+        private static readonly SemaphoreSlim s_readKeyHandle = new SemaphoreSlim(1, 1);
+
+        private static readonly SemaphoreSlim s_stdInHandle = new SemaphoreSlim(1, 1);
+
+        private Func<CancellationToken, bool> WaitForKeyAvailable;
+
+        private Func<CancellationToken, Task<bool>> WaitForKeyAvailableAsync;
 
         internal UnixConsoleOperations()
         {
             // Switch between long and short wait periods depending on if the
             // user has recently (last 5 seconds) pressed a key to avoid preventing
             // the CPU from entering low power mode.
             WaitForKeyAvailable = LongWaitForKey;
+            WaitForKeyAvailableAsync = LongWaitForKeyAsync;
+        }
+
+        internal ConsoleKeyInfo ReadKey(bool intercept, CancellationToken cancellationToken)
+        {
+            s_readKeyHandle.Wait(cancellationToken);
+
+            InputEcho.Disable();
+            try
+            {
+                while (!WaitForKeyAvailable(cancellationToken));
+            }
+            finally
+            {
+                InputEcho.Disable();
+                s_readKeyHandle.Release();
+            }
+
+            s_stdInHandle.Wait(cancellationToken);
+            try
+            {
+                return System.Console.ReadKey(intercept);
+            }
+            finally
+            {
+                s_stdInHandle.Release();
+            }
         }
 
         public async Task<ConsoleKeyInfo> ReadKeyAsync(CancellationToken cancellationToken)
         {
-            await _readKeyHandle.WaitAsync(cancellationToken);
+            await s_readKeyHandle.WaitAsync(cancellationToken);
 
             // I tried to replace this library with a call to `stty -echo`, but unfortunately
             // the library also sets up allowing backspace to trigger `Console.KeyAvailable`.
             InputEcho.Disable();
             try
             {
-                while (!await WaitForKeyAvailable(cancellationToken));
+                while (!await WaitForKeyAvailableAsync(cancellationToken));
             }
             finally
             {
                 InputEcho.Enable();
-                _readKeyHandle.Release();
+                s_readKeyHandle.Release();
             }
 
-            return System.Console.ReadKey(intercept: true);
+            await s_stdInHandle.WaitAsync(cancellationToken);
+            try
+            {
+                return System.Console.ReadKey(intercept: true);
+            }
+            finally
+            {
+                s_stdInHandle.Release();
+            }
         }
 
-        private Func<CancellationToken, Task<bool>> WaitForKeyAvailable;
+        public int GetCursorLeft()
```

Comment:
It's because the class is `internal`.  So is the interface actually.

But I do have comments on the interface itself, which is the only way it should be accessed even internally, so we should still get the documentation whenever the interface is used.